### PR TITLE
Comply with OpenTelemetry attributes specification

### PR DESCRIPTION
**Pull request by MrAlias**
Currently attributes that a users tries to add to a `Span` are ensured to be compliant with [the OpenTelemetry specification](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.0.1/specification/common/common.md#attributes) before they are added to the `Span`, otherwise they are dropped. This same logic is needed when attributes are used to create resources.

This unifies the logic to check if an attribute is valid in a new `Valid` method on the `attribute.KeyValue` type and adds this validation check to the `resource` package.

### CODE REVIEW

File: sdk/resource/resource.go [hunk 1]
```diff
@@ -43,13 +43,26 @@ var (
...
+	s, _ := attribute.NewSetWithFiltered(attrs, func(kv attribute.KeyValue) bool {
+		return kv.Valid()
+	})
+
+	// If attrs only contains invalid entries do not allocate a new resource.
+	if s.Len() == 0 {
+		return &emptyResource
+	}
+
+	return &Resource{s} //nolint
```
Comment (MrAlias) [inline sdk/resource/resource.go:65]: The exported `attribute` package API should not allow for a copylock vet error to be exposed.

File: CHANGELOG.md [hunk 2]
```diff
@@ -14,6 +14,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - A `ForceFlush` method to the `"go.opentelemetry.io/otel/sdk/trace".TracerProvider` to flush all registered `SpanProcessor`s. (#1608)
 - Added `WithDefaultSampler` and `WithSpanLimits` to tracer provider. (#1633)
 - Jaeger exporter falls back to `resource.Default`'s `service.name` if the exported Span does not have one. (#1673)
+- A `Valid` method to the `"go.opentelemetry.io/otel/attribute".KeyValue` type. (TBD)
```
Comment (XSAM) [inline CHANGELOG.md:17]:
```suggestion
- A `Valid` method to the `"go.opentelemetry.io/otel/attribute".KeyValue` type. (#1703)
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -27,6 +28,8 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
   - `trace.NewSpanContext()` can be used in conjunction with the `trace.SpanContextConfig` struct to initialize a new `SpanContext` where all values are known.
 - Renamed the `LabelSet` method of `"go.opentelemetry.io/otel/sdk/resource".Resource` to `Set`. (#1692)
 - Jaeger exporter populates Jaeger's Span Process from Resource. (#1673)
+- `"go.opentelemetry.io/otel/sdk/resource".NewWithAttributes` will now drop any invalid attributes passed. (TBD)
+- `"go.opentelemetry.io/otel/sdk/resource".StringDetector` will now error if the produced attribute is invalid. (TBD)
```
Comment (XSAM) [inline CHANGELOG.md:32]:
```suggestion
- `"go.opentelemetry.io/otel/sdk/resource".NewWithAttributes` will now drop any invalid attributes passed. (#1703)
- `"go.opentelemetry.io/otel/sdk/resource".StringDetector` will now error if the produced attribute is invalid. (#1703)
```

**Review by Aneurysm9 — APPROVED** [review]

**Review by XSAM — APPROVED** [review]
👍

**Pull request merged**